Samples/ExtendedControlAndMetadata/MediaCaptureSampleExtendedControl/CustomExtendedProperty.cpp

```diff
@@ -0,0 +1,130 @@
+//*********************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
+// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
+// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
+// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
+//
+//*********************************************************
+
+//
+// CustomExtendedProperty.cpp
+//
+
+#include "pch.h"
+#include "CustomExtendedProperty.h"
+
+using namespace CustomExtendedProperties;
+
+template<typename T>
+CustomExtendedProperty<T>::~CustomExtendedProperty()
+
+{
+    m_pExtPropHeader = nullptr;
+    m_pExtPropValue = nullptr;
+    m_cbData = 0;
+    if (m_pbData)
+    {
+        free(m_pbData);
+        m_pbData = nullptr;
+    }
+}
+
+template<typename T>
+CustomExtendedProperty<T>::CustomExtendedProperty(Platform::String^ strPropId, BYTE* pbData, DWORD cbData) :
+    m_pbData(nullptr)
+    , m_cbData(0)
+    , m_pExtPropHeader(nullptr)
+    , m_pExtPropValue(nullptr)
+{
+    m_pbData = (BYTE*)malloc(cbData);
+    if (m_pbData == nullptr)
+    {
+        return;
+    }
+    CopyMemory(m_pbData, pbData, cbData);
+
+    m_strPropId = strPropId;
+    m_cbData = cbData;
+    m_pExtPropHeader = (KSCAMERA_EXTENDEDPROP_HEADER*)m_pbData;
+    m_pExtPropHeader->Version = KSCAMERA_EXTENDEDPROP_VERSION;
+    m_pExtPropHeader->Size = sizeof(KSCAMERA_EXTENDEDPROP_HEADER) + sizeof(T);
+    m_pExtPropValue = (T*)(m_pbData + sizeof(KSCAMERA_EXTENDEDPROP_HEADER));
+}
+
+
+template<typename T>
+Platform::Array<unsigned char>^ CustomExtendedProperty<T>::GetData()
+{
+    Platform::Array<unsigned char>^ spArr = ref new Platform::Array<unsigned char>(m_pbData, m_cbData);
+    return spArr;
+}
+
+HRESULT ExtendedPropertyValueHelper::SetDeviceProperty(Windows::Media::Devices::VideoDeviceController^ vdc, Platform::String^ strPropId, CustomExtendedProperty<KSCAMERA_EXTENDEDPROP_VALUE>* pControl)
+{
+    return CExtendedPropertyHelperBase<KSCAMERA_EXTENDEDPROP_VALUE>::SetDeviceProperty(vdc, strPropId, pControl);
+}
+
+HRESULT ExtendedPropertyValueHelper::GetDeviceProperty(Windows::Media::Devices::VideoDeviceController^ vdc, Platform::String^ strPropId, CustomExtendedProperty<KSCAMERA_EXTENDEDPROP_VALUE>** ppControl)
+{
+    return CExtendedPropertyHelperBase<KSCAMERA_EXTENDEDPROP_VALUE>::GetDeviceProperty(vdc, strPropId, ppControl);
+}
+
+HRESULT ExtendedPropertyVideoProcSettingHelper::SetDeviceProperty(Windows::Media::Devices::VideoDeviceController^ vdc, Platform::String^ strPropId, CustomExtendedProperty<KSCAMERA_EXTENDEDPROP_VIDEOPROCSETTING>* pControl)
+{
+    return CExtendedPropertyHelperBase<KSCAMERA_EXTENDEDPROP_VIDEOPROCSETTING>::SetDeviceProperty(vdc, strPropId, pControl);
+}
+
+HRESULT ExtendedPropertyVideoProcSettingHelper::GetDeviceProperty(Windows::Media::Devices::VideoDeviceController^ vdc, Platform::String^ strPropId, CustomExtendedProperty<KSCAMERA_EXTENDEDPROP_VIDEOPROCSETTING>** ppControl)
```

Comment:
>ExtendedPropertyVideoProcSettingHelper [](http://example.com/codeflow?start=8&length=38)

I thought this should be throwing code?

Reply:
yeah it could but not in case of a exception that is not a Platform::Exceptiion? not sure what was the original intent, I can add a rethrow or just leave it like this

Reply:
mixing throw and hr mechanism seems confusing the callers already have a catch.  Maybe this code should be in the CameraKsProperty helper instead

Reply:
CameraKsProperty  is part of the other sample. We could coallesce both into a single sample in the future I agree to avoid c++/cx. I agree that the original intent is unclear, seems like the sample writter was not sure what he wanted to handle and covered both bases with hr and exception to no purpose